File: 6_exporters/translation.jl
#import SpineInterface
import JSON

#path = dirname(@__DIR__)
#tool = "pypsa"#"spineopt"#

#componenttooltemplate = ARGS[1]#joinpath(path,"0_component_tool_template","generic_data_format.json")#
map = ARGS[1]#joinpath(path, "6_exporters", "map_"*tool*".jl")#
input = ARGS[2]#joinpath(path, "5_built_data_and_scenarios", "casedata.json")#
output = ARGS[3]#joinpath(path, "7_tools", "input_"*tool*".json")#

include(map)

iodb = Dict{String,Any}()

# preprocess
println("Preprocessing")
map_preprocess(iodb)

# Load input data
# assume that input follows the generic energy format template
println("Load input data")
#SpineInterface.using_spinedb(input, Main)
inputdata = open(input, "r") do f
	dicttxt = read(f,String)
	return JSON.parse(dicttxt)
end
entitynames = Dict()
entitydict = Dict()
for entity in inputdata["entities"]
	if length(entity[3])>1
		entityrelation = entity[3]
	else
		entityrelation = entity[2]
	end
	entitynames[entity[2]] = entityrelation
	entitydict[entity[2]] = entity
end
parameterdict = Dict()
for parameterdefinition in inputdata["parameter_definitions"]
	entityclass = parameterdefinition[1]
	parametername = parameterdefinition[2]
	parametervalue = parameterdefinition[3]
	merge!(
		get!(parameterdict, entityclass, Dict()),
		Dict(
			parametername => parametervalue
		)
	)
end
#print(JSON.json(parameterdict,4))
entityparameterdict = Dict()
for parametervalue in inputdata["parameter_values"]
	entityname = entitynames[parametervalue[2]]
	merge!(
		get!(entityparameterdict, entityname, Dict()),
		Dict(
			parametervalue[3] => parametervalue[4]
		)
	)
end
#correct for missing values
for (entityname,entity) in entitydict
	entityclass = entity[1]
	get!(entityparameterdict, entityname, Dict())
	for (parametername,parametervalue) in parameterdict[entityclass]
		get!(entityparameterdict[entityname], parametername, parametervalue)
	end
end

# map the input data to the output data
println("Map input data to output data")
for (entityname, entityparameters) in entityparameterdict
	entities = [entityname]
	parameters = [entityparameters]
	for entityname in entitydict[entityname][3]
		push!(entities, entityname)
		push!(parameters, entityparameterdict[entityname])
	end
	getfield(Main,Symbol("map_"*entitydict[entityname][1]))(iodb,entities,parameters)
end

# postprocess
println("Postprocessing")
map_postprocess(iodb)

# write data to output
println("Writing data")
open(output, "w") do f
	JSON.print(f, iodb, 4)
end
	

File: ines_tools/translation.py
import sys
import json
from importlib.machinery import SourceFileLoader
import spinedb_api as api
from spinedb_api import purge

#path = dirname(@__DIR__)
#tool = "pypsa"#"spineopt"#

ARGS = sys.argv[1:]
mapfile = ARGS[0]
input = ARGS[1]
output = ARGS[2]

map = SourceFileLoader("map", mapfile).load_module()

iodb = {}

# preprocess
print("Preprocessing")
map.map_preprocess(iodb)

# Load input data
# assume that input follows the generic energy format template
print("Load input data")
if input.split(".")[-1] == 'json':
	with open(input) as f:
		inputdata = json.load(f)
else:
	with api.DatabaseMapping(input) as source_db:
		inputdata = api.export_data(source_db,parse_value=api.parameter_value.load_db_value)

entitydict = {}
for entity in inputdata["entities"]:
	entityrelation = ''.join(entity[1])
	entitydict[entityrelation] = entity

parameterdict = {}
for parameterdefinition in inputdata["parameter_definitions"]:
	entityclass = parameterdefinition[0]
	parametername = parameterdefinition[1]
	parametervalue = parameterdefinition[2]
	if entityclass not in parameterdict:
		parameterdict[entityclass] = {}
	parameterdict[entityclass][parametername] = parametervalue

entityparameterdict = {}
for parametervalue in inputdata["parameter_values"]:
	entityname = ''.join(parametervalue[1])
	if entityname not in entityparameterdict:
		entityparameterdict[entityname] = {}
	entityparameterdict[entityname][parametervalue[2]] = parametervalue[3]

#correct for missing values
for (entityname,entity) in entitydict.items():
	entityclass = entity[0]
	if entityname not in entityparameterdict:
		entityparameterdict[entityname] = {}
	for (parametername,parametervalue) in parameterdict[entityclass].items():
		if parametername not in entityparameterdict[entityname]:
			entityparameterdict[entityname][parametername] = parametervalue

# map the input data to the output data
print("Map input data to output data")
for (entityname, entityparameters) in entityparameterdict.items():
	# an entity (or relation between entities) and its parameters
	entities = [entitydict[entityname][1]]
	parameters = [entityparameters]
	if type(entitydict[entityname][1]) != str:
		# list or tuple
  		# in other words, if the entity describes a relation
		# we add the related entities as they may be needed in the translation
		for entityrelation in entitydict[entityname][1]:
			entities.append(entityrelation)
			parameters.append(entityparameterdict[entityrelation])
	getattr(map,"map_"+entitydict[entityname][0])(iodb,entities,parameters)

# postprocess
print("Postprocessing")
map.map_postprocess(iodb)

# write data to output
print("Writing data")
if output.split(".")[-1] == 'json':
	with open(output, 'w') as f:
		json.dump(iodb, f, indent=4)
else:
	with api.DatabaseMapping(output) as target_db:
		purge.purge(target_db, purge_settings=None)
		api.import_data(target_db,**iodb)
		target_db.refresh_session()
		target_db.commit_session("Import data")